[ao] Added OutlierDetector observer insert implementation

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80937
* __->__ #80880

Summary: This adds the implementation for observer insertion point
selection for the OutlierDetector. For this detector, the insertion
points are to insert a ModelReportObserver before any leaf level module
to study the distribution of data that passes into the module to detect
outliers.

This commit contains the implementation of the observer insertion as
well as the relavent test case. Some code from the
InputWeightEqualization was abstracted and made more modular so the same
helper function could be used for multiple outlier class tests.

As a part of the work for this, there was testing done to determine what
a good default ratio threshold and reference percentile would be, and
the work to determine this (based on a normal distribution) was then
analyzed to find good paramters.

We still want to keep thresholds and reference percentile as something
the user can input because these were based on a normal distribution,
and it can definately vary depending on the type of data a user has.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Added OutlierDetector observer insert implementation

Summary: This adds the implementation for observer insertion point
selection for the OutlierDetector. For this detector, the insertion
points are to insert a ModelReportObserver before any leaf level module
to study the distribution of data that passes into the module to detect
outliers.

This commit contains the implementation of the observer insertion as
well as the relavent test case. Some code from the
InputWeightEqualization was abstracted and made more modular so the same
helper function could be used for multiple outlier class tests.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1319,5 +1304,92 @@ def test_input_weight_equalization_report_gen_empty(self):
+
+    # if they passed in fusion paramter, make sure to test that
+    if fused:
+        model = torch.quantization.fuse_modules(model, [['conv', 'relu']])
```
this does not look like a general helper function, should we attach this to the model definition (LargeBatchModel)?

vspenubarthi replied:
it is general as it is used for multiple different models across multiple different test classes. For example, lines 1153 and 1154 (in `test_input_weight_equalization_determine_points`) also make use of this to test a fusion vs. non-fusion case for a different model class.

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1319,5 +1304,92 @@ def test_input_weight_equalization_report_gen_empty(self):
+
+            # get tst model and callibrate
+            non_fused = self._get_prepped_for_calibration_model(self.LargeBatchModel(), detector_set)
+            fused = self._get_prepped_for_calibration_model(self.LargeBatchModel(), detector_set)
```
is there a missing `fused = True` here?

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1319,5 +1304,92 @@ def test_input_weight_equalization_report_gen_empty(self):
+    def _get_prepped_for_calibration_model(self, model, detector_set):
+        r"""Returns a model that has been prepared for callibration and corresponding model_report"""
+        # call the general helper function to callibrate
+        example_input = torch.arange(128 * 128 * 128).reshape((1, 128, 128, 128))
```
this is also model specific, maybe we should attach this to the model itself?

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1319,5 +1304,92 @@ def test_input_weight_equalization_report_gen_empty(self):
+                mods_to_check = set([nn.Linear, nn.Conv2d, nn.ReLU])
+
+                # get the set of all nodes in the graph their fqns
+                node_fqns = set([node.target for node in prepared_for_callibrate_model.graph.nodes])
```
do we need to filter with node.op == "call_module"?

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1319,5 +1304,92 @@ def test_input_weight_equalization_report_gen_empty(self):
+                number_of_obs_found = 0
+                obs_name_to_find = InputWeightEqualizationDetector.DEFAULT_PRE_OBSERVER_NAME
+
+                for node in prepared_for_callibrate_model.graph.nodes:
```
should we be using `node_fqns` here? looks like it's not used

vspenubarthi replied:
was a relic of some old code that was used here, node_fqns is removed

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1319,5 +1304,92 @@ def test_input_weight_equalization_report_gen_empty(self):
+
+                for node in prepared_for_callibrate_model.graph.nodes:
+                    # if the obs name is inside the target, we found an observer
+                    if obs_name_to_find in str(node.target):
+                        number_of_obs_found += 1
```
nit: this can be a list comprehension as well

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -10,6 +10,13 @@ class ModelReportObserver(ObserverBase):
     Dynamic or Static Quantization is more appropriate for their model given the general
     distributions of their data.
 
+    Args:
```
should we have different observers for different tasks?

vspenubarthi replied:
That was a design decision @HDCharles and I considered, but we thought it would make more sense to keep all the functionality in one observer because:
1.) it would be easier to only deal with the adding and removal of observers if we were only looking for one
2.) if the user wanted to perform a further deep dive into the extra collected statistics, it made more sense for them to only have to look for one type of observer
3.) some of the collected statistics by the observer are used multiple times in each of the observer "data collection flows" therefore it also helps cut down on redundant code

andrewor14 replied:
I also think it's a bit weird for observers associated with a completely unrelated detector to have a `comp_percentile` argument when this is only used to detect outliers. Maybe we should discuss this more offline after this PR is merged.

andrewor14 replied:
Separating the observers per task would also allow you to structure the code more cleanly by breaking down this giant class (right now we have to keep adding fields/methods to it every time we have a new detector)

vspenubarthi committed:
Update on "[ao] Added OutlierDetector observer insert implementation"


Summary: This adds the implementation for observer insertion point
selection for the OutlierDetector. For this detector, the insertion
points are to insert a ModelReportObserver before any leaf level module
to study the distribution of data that passes into the module to detect
outliers.

This commit contains the implementation of the observer insertion as
well as the relavent test case. Some code from the
InputWeightEqualization was abstracted and made more modular so the same
helper function could be used for multiple outlier class tests.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1134,26 +1138,10 @@ def forward(self, x):
-        prepared_for_callibrate_model = model_report.prepare_detailed_calibration(model_prep)
-
-        return (prepared_for_callibrate_model, model_report)
+        return _get_prepped_for_calibration_model_helper(model, detector_set, example_input, fused)
```
should we make `example_input`/`example_inputs` a method for `model` as well?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -931,26 +931,57 @@ class OutlierDetector(DetectorBase):
+        # check if the module has any children and isn't observer
+        if insert:
+            num_children = len(list(module.children()))
+            return num_children == 0 and not isinstance(module, ObserverBase)
```
we can use https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/quantize.py#L47 for `not isinstance(module, ObserverBase)`

jerryzh168 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -931,26 +931,57 @@ class OutlierDetector(DetectorBase):
+
+        Args
+            module: The module to check and ensure is supported
+            insert: True if this is check for observer insertion, false if for report gen
```
this boolean flag feels a bit weird, maybe worth considering just have two `is_supported` method, one for observer insertion, one for report gen. I'm not sure how these two things are related

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -113,6 +136,75 @@ def _calculate_min_max_stats(self, x_copy):
+        cent_quartile = torch.quantile(y, quantiles_to_find[1], dim=self.ch_axis, interpolation="lower")
+
+        # if any of the channels have 0s, we ignore that channel for this calculation
+        ch_no_zeros: torch.Tensor = (comp_quantile != torch.tensor([0])) | (cent_quartile != torch.tensor([0]))
```
can we have a more descriptive name? like `any_zero_quantile_value`, also should `comp_quantile` and `cent_quantile` be vectors?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -113,6 +136,75 @@ def _calculate_min_max_stats(self, x_copy):
+
+        # find the quantiles
+        comp_quantile = torch.quantile(y, quantiles_to_find[0], dim=self.ch_axis, interpolation="lower")
+        cent_quartile = torch.quantile(y, quantiles_to_find[1], dim=self.ch_axis, interpolation="lower")
```
is `cent_quartile` a typo?

vspenubarthi commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -931,26 +931,57 @@ class OutlierDetector(DetectorBase):
     Args:
-        ratio_threshold (float): The threshold for p_r to determine if there are outliers in activations
-            Should be between 0 and 1 (both non-inclusive)
+        ratio_threshold (float, optional): The threshold for p_r to determine if there are outliers in activations
```
yup, i updated the description to reflect that and add more on their relation

vspenubarthi committed:
Update on "[ao] Added OutlierDetector observer insert implementation"


Summary: This adds the implementation for observer insertion point
selection for the OutlierDetector. For this detector, the insertion
points are to insert a ModelReportObserver before any leaf level module
to study the distribution of data that passes into the module to detect
outliers.

This commit contains the implementation of the observer insertion as
well as the relavent test case. Some code from the
InputWeightEqualization was abstracted and made more modular so the same
helper function could be used for multiple outlier class tests.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -931,26 +932,55 @@ class OutlierDetector(DetectorBase):
+            Default: 3.5
         reference_percentile (float, optional): The denominator to find the relative scale of the 100th percentile
+            Should be between 0 and 1
+            Default: 0.975
```
Also is there a good `ratio_threshold` for each `reference_percentile`, assuming the data follows a normal distribution? If so I think this would make it easier to provide a more appropriate default for `ratio_threshold` (as opposed to just 3.5) for different `reference_percentile` values

vspenubarthi replied:
https://www.internalfb.com/intern/anp/view/?id=2169566 there was work done in this notebook to determine some good values and their respective thresholds. These were picked based on the experiment done in that notebook as to what the ratio would look like in a normal distribution.

I change it so that the user can select something from a fixed set of values, and the values have already had their respective threshold value calculated.

vspenubarthi replied:
I also think I still want it to be a value the user can pass in, because depending on their own data distribution, what counts as an outlier or what percentile they want to look at can change drastically. Furthermore, because the ratio_threshold was set based on a normal distribution, if their data is inherently more of a two-peaks distribution, then this just wouldn't really work, which is why i thought allowing as much flexibility as possible was appropriate.

However, I see your point and I definitely think this is something to revisit later if we find that almost everything we encounter will just be a normal distribution and it will be better to just give users the different options to choose from there.

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -931,26 +932,55 @@ class OutlierDetector(DetectorBase):
+        This threshold was calculated based on the ratio of the percentiles in a normal distribution
 
     * :attr:`reference_percentile`: The denominator of the top fraction to find the relative scale of the 100th percentile
+        Should be between 0 and 1
```
We should add some checks to throw an exception if this is not satisfied

vspenubarthi commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -931,26 +932,55 @@ class OutlierDetector(DetectorBase):
-        ratio_threshold (float): The threshold for p_r to determine if there are outliers in activations
-            Should be between 0 and 1 (both non-inclusive)
+        ratio_threshold (float, optional): The threshold for p_r to determine if there are outliers in activations
+            Default: 3.5
```
https://www.internalfb.com/intern/anp/view/?id=2169566 there was work done in this notebook to determine some good values and their respective thresholds. These were picked based on the experiment done in that notebook as to what the ratio would look like in a normal distribution.

vspenubarthi replied:
I'll include a link to this notebook in the commit message, thanks for pointing this out

vspenubarthi replied:
Added to the commit message to explain reasoning better instead since this is already linked here.

vspenubarthi committed:
Update on "[ao] Added OutlierDetector observer insert implementation"


Summary: This adds the implementation for observer insertion point
selection for the OutlierDetector. For this detector, the insertion
points are to insert a ModelReportObserver before any leaf level module
to study the distribution of data that passes into the module to detect
outliers.

This commit contains the implementation of the observer insertion as
well as the relavent test case. Some code from the
InputWeightEqualization was abstracted and made more modular so the same
helper function could be used for multiple outlier class tests.

As a part of the work for this, there was testing done to determine what
a good default ratio threshold and reference percentile would be, and
the work to determine this (based on a normal distribution) was then
analyzed to find good paramters.

We still want to keep thresholds and reference percentile as something
the user can input because these were based on a normal distribution,
and it can definately vary depending on the type of data a user has.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added OutlierDetector observer insert implementation"


Summary: This adds the implementation for observer insertion point
selection for the OutlierDetector. For this detector, the insertion
points are to insert a ModelReportObserver before any leaf level module
to study the distribution of data that passes into the module to detect
outliers.

This commit contains the implementation of the observer insertion as
well as the relavent test case. Some code from the
InputWeightEqualization was abstracted and made more modular so the same
helper function could be used for multiple outlier class tests.

As a part of the work for this, there was testing done to determine what
a good default ratio threshold and reference percentile would be, and
the work to determine this (based on a normal distribution) was then
analyzed to find good paramters.

We still want to keep thresholds and reference percentile as something
the user can input because these were based on a normal distribution,
and it can definately vary depending on the type of data a user has.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
